doc/format-directory.md

```diff
@@ -0,0 +1,178 @@
+metadata file.
+
+In particular, the snapshot metadata file tells us exactly which other LSM run
+files to expect. This is important for detecting filesystem corruption.
+
+# LSM run file layout
+
+Each completed LSM run has five files:
+
+ 1. `${n}.keyops`: the sorted run of key/operation pairs
+ 2. `${n}.blobs`:  the blob values associated with the key/operations
+ 3. `${n}.filter`: a Bloom filter of all the keys in the run
+ 4. `${n}.index`:  an index from keys to disk page numbers
+ 5. `${n}.checksum`: a file listing the crc32c checksums of the other files
+
+The format of the main four files is covered in `format-run.md`.
+
+Each in-progress LSM run has only the main files, but no checksum file, and the
+main files will be in some intermediate state of construction.
+
+# Checksum files
+
+Each `.checksum` file lists the CRC-32C (Castagnoli) of other files.
+
+The file uses the BSD-style checksum format (e.g. as produced by tools like
+`md5sum --tag`), with the algorithm name "CRC32C". This format is text,
+one line per file, using hexedecimal for the 32bit output.
+
+Checksum files are used for each LSM run, and for the snapshot metadata.
+```
+CRC32C (keyops) = fd040004
+CRC32C (blobs) = 5a3b820c
+CRC32C (filter) = 6653e178
+CRC32C (index) = f4ec6724
+```
+```
+CRC32C (snapshot) = 87972d7f
+```
+Note that the checksum file for LSM runs _does not_ include the full name of
+each file, such as `0.keyops`. It excludes the number from the file name. The
+number is instead implicit from the name of the checksum file, e.g.
+`0.checksum` contains checksums for `0.*` files.
+
+The reason for this kind of "relative" naming, is because when the LSM runs
+are moved between active and snapshot, the number of the LSM run will change,
+and by using relative names, we avoid having to rewrite the checksum files,
+and can instead just hard link them.
+
+# Verifying snapshots
+
+The checksum of the snapshot metadata must be verified.
+
+The snapshot metadata file then determines the number of LSM runs within the
+snapshot. Each numbered run within the snapshot must have a checksum file,
+which must list the other files.
+
+The checksum of each LSM run file must be verified.
+
+If verfification passes, this ensures all the files are present and have their
+expected content. The use of CRCs protects against accidental corruption, not
+deliberate corruption.
+
+# Saving and restoring snapshots using hard links
+
+Saving a snapshot involves a few steps:
+
+ 1. creating a _new_ directory for the snapshot
+ 2. writing out the snapshot metadata file and checksum
+ 3. hard linking all the LSM run files into the snapshot directory
+ 4. for durability, `fsync`:
+   - all the LSM run files, including checksum files
+   - metadata file and metadata checksum file
+   - the snapshot directory itself
+
+All five files for each LSM (including the checksum file) are hard-linked into
+the snapshot directory under new number. The number of the file changes
```

Comment:
Is it necessary to renumber them if we are probably renumbering them anyway once the snapshot is restored? That is, assuming that the live directory might contain any number of live handles

Reply:
No, not strictly necessary, but it seems cleaner.